### PR TITLE
Add Sprig Templating

### LINKED ISSUE #540: Enhance templating language with sprig
**Describe the feature:**

Review and add sprig functions to goss templating.

https://github.com/Masterminds/sprig

### PR DESCRIPTION
**Pull request by pedroMMM**
##### Checklist

- [x] `make test-all` (UNIX) passes. CI will also test this
- [x] unit and/or integration tests are included (if applicable)
- [x] documentation is changed or added (if applicable)

### Description of change

Simply add all Sprig templating functions and start a discussion.

Closes #540

### CODE REVIEW

**Comment by pedroMMM** [comment]
@aelsabbahy I haven't much time but this is a quick win that I can pull off.

Any particular functions you were hoping to showcase in the documentation? 

I am also _hoping_ to get an afternoon or so to play with the integration testing expansion that we have talked since this is one of these PRs that unit tests don't make much sense.

**Comment by aelsabbahy** [comment]
Funny, I was actually working on this last night.

I'll throw a PR out there tomorrow or the day after tagging you. Honestly, going through the sprig functions not all of them make sense for goss.

**Comment by pedroMMM** [comment]
image

My bad I missed the assignment 😆! 

> Honestly, going through the sprig functions not all of them make sense for goss.

Well I thought about that as well like looking at the API... I thought about importing the Sprg functions one by one instead of pulling them all. At the same time that is more work upfront and to maintain, so just having all even if they don't make sense in the Goss context.

**Comment by pedroMMM** [comment]
@aelsabbahy I will wait for your PR. Let me know if you want me to jump onto this instead.

**Comment by aelsabbahy** [comment]
Sorry for the delay in this. So my initial thoughts was to support a subset of the [HermeticTxtFuncMap()](https://github.com/Masterminds/sprig/blob/master/functions.go#L29)

But honestly, at this point I'm thinking your solution is just much easier.

We should have an example file and an example_rendered version for documentation.

**Comment by pedroMMM** [comment]
> Sorry for the delay in this. So my initial thoughts was to support a subset of the [HermeticTxtFuncMap()](https://github.com/Masterminds/sprig/blob/master/functions.go#L29)
> 
> But honestly, at this point I'm thinking your solution is just much easier.

Ups sorry, I shouldn't have closed it. 

> We should have an example file and an example_rendered version for documentation.

Ok I will get the documentation up and running and add an example based on this `hello` example plus a link to their documentation as well.

**Comment by pedroMMM** [comment]
@aelsabbahy The documentation is been added. Please let me know what you think when you have time. 

I explained the load order and that the custom Goss functions will overwrite the Sprig functions.

File: docs/manual.md [hunk 1]
```diff
@@ -871,20 +871,24 @@ For more information see:
...
 
-Available functions beyond text/template [built-in functions](https://golang.org/pkg/text/template/#hdr-Functions):
-* `mkSlice "ARG1" "ARG2"` - Retuns a slice of all the arguments. See examples below for usage.
-* `getEnv "var" ["default"]` - A more forgiving env var lookup. If key is missing either "" or default (if provided) is returned.
-* `readFile "fileName"` - Reads file content into a string, trims whitespace. Useful when a file contains a token.
-  * **NOTE:** Goss will error out during during the parsing phase if the file does not exist, no tests will be executed.
-* `regexMatch "(some)?reg[eE]xp"` - Tests the piped input against the regular expression argument.
-* `toLower` - Changes piped input to lowercase
-* `toUpper` - Changes piped input to UPPERCASE
+Available functions **in load order**:
```
Comment (aelsabbahy) [inline docs/manual.md]: Not sure if "in load order" is clear to new users.

I wonder if there's a better way to word.

Comment (pedroMMM) [reply docs/manual.md]: That is fair, I struggled with the wording here. I tweak it by removing the `in load order` bit and explained at right after the list that some of Custom Goss Functions might overwrite the Sprig Functions and we are keeping the custom functions for backward compatibility.

File: docs/manual.md [hunk 2]
```diff
@@ -895,22 +895,26 @@ For more information see:
...
-* `getEnv "var" ["default"]` - A more forgiving env var lookup. If key is missing either "" or default (if provided) is returned.
-* `readFile "fileName"` - Reads file content into a string, trims whitespace. Useful when a file contains a token.
-  * **NOTE:** Goss will error out during during the parsing phase if the file does not exist, no tests will be executed.
-* `regexMatch "(some)?reg[eE]xp"` - Tests the piped input against the regular expression argument.
-* `toLower` - Changes piped input to lowercase
-* `toUpper` - Changes piped input to UPPERCASE
+Available functions:
+* [built-in text/template functions](https://golang.org/pkg/text/template/#hdr-Functions)
+* [Sprig functions](https://masterminds.github.io/sprig/)
+* Custom functions ():
```
Comment (aelsabbahy) [inline docs/manual.md]:
```suggestion
* Custom functions:
```

File: docs/manual.md [hunk 3]
```diff
@@ -895,22 +895,26 @@ For more information see:
...
+  * `mkSlice "ARG1" "ARG2"` - Returns a slice of all the arguments. See examples below for usage.
+  * `getEnv "var" ["default"]` - A more forgiving env var lookup. If key is missing either "" or default (if provided) is returned.
+  * `readFile "fileName"` - Reads file content into a string, trims whitespace. Useful when a file contains a token.
+    * **NOTE:** Goss will error out during during the parsing phase if the file does not exist, no tests will be executed.
+  * `regexMatch "(some)?reg[eE]xp"` - Tests the piped input against the regular expression argument.
+  * `toLower` - Changes piped input to lowercase
+  * `toUpper` - Changes piped input to UPPERCASE
 
 **NOTE:** gossfiles containing text/template `{{}}` controls will no longer work with `goss add/autoadd`. One way to get around this is to split your template and static goss files and use [gossfile](#gossfile) to import.
+**NOTE:** Some of Sprig functions have the same name as the older Custom Goss functions. The Sprig are been overwritten by by the custom functions for backwards compatibility.
```
Comment (aelsabbahy) [inline docs/manual.md]:
```suggestion
**NOTE:** Some of Sprig functions have the same name as the older Custom Goss functions. The Sprig functions are overwritten by the custom functions for backwards compatibility.
```

**Comment by donmstewart** [comment]
Great news as I too was about to look at adding Sprig but you've already done it :)

**Comment by aelsabbahy** [comment]
Yup, seems the next release will have a nice set of features implemented.

Thanks for all your contributions.

**Comment by aelsabbahy** [comment]
I'm only waiting so I can merge + cut a release.. Hopefully over this week the other outstanding issues are good to go and I can just merge/cut a release with everything in it.

If no progress on other PRs, I'll merge this one and just release this. Many thanks for the contribution.

**Comment by pedroMMM** [comment]
Thank you for all your hard work and kind words. 👍

**Comment by aelsabbahy** [comment]
Holding off on this, since I want to group up the v0.4.0 milestone stuff together. Hoping I have time to iron out the io.reader gomega matcher before then.

**Pull request merged**